firebase-config/src/main/java/com/google/firebase/remoteconfig/internal/ConfigSharedPrefsClient.java

```diff
@@ -234,21 +253,93 @@ void setLastTemplateVersion(long templateVersion) {
   BackoffMetadata getBackoffMetadata() {
     synchronized (backoffMetadataLock) {
       return new BackoffMetadata(
-          frcMetadata.getInt(NUM_FAILED_FETCHES_KEY, NO_FAILED_FETCHES),
-          new Date(frcMetadata.getLong(BACKOFF_END_TIME_IN_MILLIS_KEY, NO_BACKOFF_TIME_IN_MILLIS)));
+          frcSharedPrefs.getInt(NUM_FAILED_FETCHES_KEY, NO_FAILED_FETCHES),
+          new Date(
+              frcSharedPrefs.getLong(BACKOFF_END_TIME_IN_MILLIS_KEY, NO_BACKOFF_TIME_IN_MILLIS)));
     }
   }
 
   void setBackoffMetadata(int numFailedFetches, Date backoffEndTime) {
     synchronized (backoffMetadataLock) {
-      frcMetadata
+      frcSharedPrefs
           .edit()
           .putInt(NUM_FAILED_FETCHES_KEY, numFailedFetches)
           .putLong(BACKOFF_END_TIME_IN_MILLIS_KEY, backoffEndTime.getTime())
           .apply();
     }
   }
 
+  public void setCustomSignals(Map<String, String> newCustomSignals) {
+    synchronized (customSignalsLock) {
+      // Retrieve existing custom signals
+      Map<String, String> existingCustomSignals = getCustomSignals();
+      // Tracks whether the custom signals have been modified.
+      boolean modified = false;
+
+      for (Map.Entry<String, String> entry : newCustomSignals.entrySet()) {
+        String key = entry.getKey();
+        String value = entry.getValue();
+
+        // Validate key and value length
+        if (key.length() > CUSTOM_SIGNALS_MAX_KEY_LENGTH
```

Comment:
Since you already have an object wrapping these values, i.e. `CustomSignal`, making these checks at that level would move the error point closer to where it's caused (the actual line adding the signal that violates the constrains) and out of a synchronized block

Reply:
Initially, we considered this approach. However, we were advised against throwing exceptions, and logging errors in the builder might seem awkward to developers, since we might skip signals breaching limits or end up returning an empty map of signals. We want the setting of custom signals to be atomic—either all signals are set, or none are. Handling validation errors within setCustomSignals achieves that objective. What do you think?

Reply:
If developers expect setting custom signals to work as you mention, and you've been advise to follow that route, sgtm.